app/src/main/java/free/rm/skytube/businessobjects/YouTube/Tasks/GetBulkSubscriptionVideosTask.java

```diff
@@ -51,43 +43,65 @@ public Progress(YouTubeChannel dbChannel, int newVideos) {
-                    YouTubeVideo details;
-                    try {
-                        details = NewPipeService.get().getDetails(vid.getId());
-                        details.setChannel(vid.getChannel());
-                        detailedList.add(details);
-                    } catch (ExtractionException | IOException e) {
-                        Logger.e(this, "Error during parsing video page for " + vid.getId() + ",msg:" + e.getMessage(), e);
-                        e.printStackTrace();
-                    }
-                }
-            }
-            publishProgress(new Progress(dbChannel, newVideos.size()));
+    protected Void doInBackground(Void... params) {
+        AtomicBoolean changed = new AtomicBoolean(false);
+
+        for(final YouTubeChannel channel : channels) {
+            tasks.add(new GetBulkChannelVideosTask(channel)
+                    .setGetChannelVideosTaskInterface(videos -> {
+                        numTasksFinished++;
+                        boolean videosDeleted = false;
+                        int numberOfVideos = videos != null ? videos.size() : 0;
+                        if (numberOfVideos > 0) {
+                            changed.compareAndSet(false, true);
+                        }
+                        if(numTasksFinished < numTasksLeft) {
+                            if(tasks.size() > 0) {
+                                // More channels to fetch videos from
+                                tasks.get(0).executeInParallel();
+                                tasks.remove(0);
+                            }
+                            publishProgress(new Progress(channel, videos.size()));
+                        } else {
+                            videosDeleted = SubscriptionsDb.getSubscriptionsDb().trimSubscriptionVideos();
+
+                            // All channels have finished querying. Update the last time this refresh was done.
+                            GetSubscriptionVideosTask.updateFeedsLastUpdateTime();
+
+                            publishProgress(new Progress(channel, videos.size()));
+
+                            if (listener != null) {
+                                listener.onAllChannelVideosFetched(videosDeleted || changed.get());
+                            }
+                        }
+                    })
+
+            );
+
         }
-        subscriptionsDb.saveVideos(detailedList);
-        GetSubscriptionVideosTask.updateFeedsLastUpdateTime(System.currentTimeMillis());
-        return !detailedList.isEmpty();
-    }
 
-    @Override
-    protected void onPostExecute(Boolean changed) {
-        listener.onAllChannelVideosFetched(changed);
+        numTasksLeft = tasks.size();
+
+        int numToStart = tasks.size() >= 4 ? 4 : tasks.size();
+
+        // Start fetching videos for up to 4 channels simultaneously.
+        for(int i=0;i<numToStart;i++) {
+            tasks.get(0).executeInParallel();
+            tasks.remove(0);
+        }
+
```

Comment:
I know that the same code pattern is used in GetSubscriptionVideosTask - but this is too complex, hard to follow, and contains a couple of concurrency problems which could cause very strange behaviors - for example tasks execute multiple times, or tasks skipped entirely, or a locked UI (where the ui still thinks, that there are ongoing background tasks....)
 I would rather rewrite it with using proper java constructs - with a Semaphore and a CountDownLatch, with pseudo-code:

```
CountDownLatch countDown = new CountDownLatch(channels.size());
Semaphore semaphore = new Semaphore(4);
for (YouTubeChannel channel : channels) {
    semaphore.acquire();
    new GetBulkChannelVideosTask(countDown, semaphore).executeInParallel();
// in the GetBulkChannelVideosTask.doInBackground method:
///     ... actual network access...
///       semaphore.release();
///       countDown.countDown();
}

countDown.await();
// final result calculation..

}
```